File: sdks/python/speechswitch/http.py
"""Transport contracts and byte-native streaming for handwritten providers."""

import asyncio
from collections.abc import AsyncIterator, Mapping
from contextlib import asynccontextmanager
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class HttpRequest:
    method: str
    url: str
    headers: Mapping[str, str]
    body: bytes


class HttpBody(Protocol):
    def __aiter__(self) -> AsyncIterator[bytes]: ...

    async def __anext__(self) -> bytes: ...

    async def aclose(self) -> None:
        """Release the response, including after a canceled or failed read."""
        ...


@dataclass(frozen=True, slots=True)
class HttpResponse:
    status: int
    headers: Mapping[str, str]
    body: HttpBody


class HttpTransport(Protocol):
    async def send(self, request: HttpRequest) -> HttpResponse:
        """Return at headers, without buffering audio or blocking the event loop.

        Cancellation before return must release the request and any response.
        Never forward credentials across origins when following redirects.
        """
        ...


class HttpStatusError(Exception):
    def __init__(self, status: int) -> None:
        self.status = status
        super().__init__(f"HTTP request failed with status {status}")


class AudioStream:
    """Single-reader byte stream. Use open_audio's context for deterministic cleanup."""

    def __init__(self, body: HttpBody) -> None:
        self._body: HttpBody | None = body

    def __aiter__(self) -> AsyncIterator[bytes]:
        return self

    async def __anext__(self) -> bytes:
        try:
            while self._body is not None:
                chunk = await self._body.__anext__()
                if chunk:
                    return chunk
                await asyncio.sleep(0)
        except StopAsyncIteration:
            await self.aclose()
            raise
        except BaseException:
            # Preserve the read error (especially CancelledError), even if
            # the transport also reports an error while releasing its body.
            try:
                await self.aclose()
            except Exception:
                pass
            raise
        raise StopAsyncIteration

    async def aclose(self) -> None:
        # No concurrent reads/closes: cancellation is delivered to the reading
        # task, which releases the body before propagating CancelledError.
        body, self._body = self._body, None
        if body is not None:
            await body.aclose()


@asynccontextmanager
async def open_audio(transport: HttpTransport, request: HttpRequest) -> AsyncIterator[AudioStream]:
    """Open byte-native audio; codecs for framed responses belong to providers.

    Always use ``async with``. Breaking an async for loop alone does not close
    its iterator. Non-2xx bodies are closed without reading or exposing them.
    """
    response = await transport.send(request)
    stream = AudioStream(response.body)
    try:
        if not 200 <= response.status < 300:
            raise HttpStatusError(response.status)
        yield stream
    except BaseException:
        try:
            await stream.aclose()
        except Exception:
            pass
        raise
    finally:
        await stream.aclose()

File: sdks/rust/src/http.rs
//! Streaming HTTP ownership shared by handwritten provider adapters.
//! Wire request construction and response codecs belong to the provider.

use crate::runtime::{InputStream, StreamingInput};
use std::{
    error::Error,
    fmt,
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

pub type TransportError = Box<dyn Error + Send + Sync>;

pub struct HttpRequest {
    pub method: String,
    pub url: String,
    pub headers: Vec<(String, String)>,
    pub body: Vec<u8>,
}

pub struct HttpResponse {
    pub status: u16,
    pub headers: Vec<(String, String)>,
    pub body: StreamingInput<Vec<u8>>,
}

/// The transport owns HTTP/TLS. Dropping a pending send future must cancel the
/// request; dropping its response body must stop reading and release resources.
/// Do not forward credentials on cross-origin redirects. Neither sending nor
/// polling may block the executor, or buffer the complete response before returning.
pub trait HttpTransport: Send + Sync {
    fn send(
        &self,
        request: HttpRequest,
    ) -> Pin<Box<dyn Future<Output = Result<HttpResponse, TransportError>> + Send + '_>>;
}

#[derive(Debug, PartialEq, Eq)]
pub struct HttpStatusError {
    pub status: u16,
}

impl fmt::Display for HttpStatusError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "HTTP request failed with status {}", self.status)
    }
}
impl Error for HttpStatusError {}

/// Open a byte-native response. Framed JSON/SSE responses need a provider codec,
/// not this helper. Error bodies are closed without being buffered or exposed.
pub async fn open_audio(
    transport: &dyn HttpTransport,
    request: HttpRequest,
) -> Result<AudioStream, TransportError> {
    let response = transport.send(request).await?;
    if !(200..300).contains(&response.status) {
        return Err(Box::new(HttpStatusError {
            status: response.status,
        }));
    }
    Ok(AudioStream {
        body: Some(response.body),
    })
}

/// Pulling reads at most one transport chunk. Drop cancels unfinished audio.
/// EOF and read failure immediately release the body; an error is emitted once.
pub struct AudioStream {
    body: Option<StreamingInput<Vec<u8>>>,
}

impl InputStream<Vec<u8>> for AudioStream {
    fn poll_next(
        self: Pin<&mut Self>,
        context: &mut Context<'_>,
    ) -> Poll<Option<Result<Vec<u8>, TransportError>>> {
        let stream = self.get_mut();
        let Some(body) = &mut stream.body else {
            return Poll::Ready(None);
        };
        match body.as_mut().poll_next(context) {
            Poll::Ready(Some(Ok(bytes))) if bytes.is_empty() => {
                // Empty chunks aren't EOF. Yield to the executor instead of
                // spinning indefinitely on an immediately-ready producer.
                context.waker().wake_by_ref();
                Poll::Pending
            }
            result @ (Poll::Ready(None) | Poll::Ready(Some(Err(_)))) => {
                stream.body = None;
                result
            }
            result => result,
        }
    }
}
